File: day_1/day_1.py
import sys
import re
import math
import itertools
import functools
import collections


day = 1


def read_input(day) -> str:
    with open("./input_day_{day}.txt".format(day=day), "r") as f:
        input_data = f.read()
    return input_data


def parse_input():
    input_data = read_input(day).splitlines()
    left = []
    right = []
    for line in input_data:
        numbers = line.split()
        left.append(int(numbers[0]))
        right.append(int(numbers[1]))
    return left, right


left, right = parse_input()
left.sort()
right.sort()

print(sum(abs(a - b) for a, b in zip(left, right)))

freq = collections.Counter(right)
print(sum(a * freq[a] for a in left if a in freq))

File: day_1/day_1.go
package main

import (
	"fmt"
	"log"
	"os"
	"slices"
	"strconv"
	"strings"
)

var day = 1

func readInput() string {
	input_file_path := fmt.Sprintf("input_day_%d.txt", day)
	f, err := os.ReadFile(input_file_path)
	if err != nil {
		log.Fatalf("error reading file: %v", err)
	}
	return string(f)
}

func parseInput(input string) ([]int, []int) {
	var left, right []int
	lines := strings.Split(input, "\r\n")
	for _, line := range lines {
		if line == "" {
			continue
		}
		numbers := strings.SplitN(line, " ", 2)
		num1, _ := strconv.Atoi(strings.TrimSpace(numbers[0]))
		num2, _ := strconv.Atoi(strings.TrimSpace(numbers[1]))
		left = append(left, num1)
		right = append(right, num2)
	}
	return left, right
}

func main() {
	left, right := parseInput(readInput())
	slices.Sort(left)
	slices.Sort(right)
	sum := 0
	for i, l := range left {
		if l > right[i] {
			sum += l - right[i]
		} else {
			sum += right[i] - l
		}
	}

	rightCount := map[int]int{}
	for _, r := range right {
		rightCount[r]++
	}

	score := 0
	for _, l := range left {
		score += (l * rightCount[l])
	}

	fmt.Println(sum)
	fmt.Println(score)
}
